src/acl/Ip.cc

```diff
@@ -252,12 +252,80 @@ acl_ip_data::FactoryParse(const char *t)
         return q;
     }
 
-    /* Special ACL RHS "ipv6" matches IPv6-Unicast Internet */
+    /* Special ACL RHS "ipv6" matches IPv6 Internet */
```

Comment:
Matches addresses which are exclusively IPv6 (or handled as such by Squid, as a few of the newer mapped ranges are [for now]).
FWIW, I removed the "Unicast" from this comment only because the addition of "Multicast" ranges makes that technically invalid. All ranges except that one should be IPv6-Unicast.

Reply:
Can you detail "exclusively IPv6" more, please? For example, can we say that "ipv6" should match any Ip::Address object X for which X.isIPv6() returns true? If yes, then "exclusively IPv6" means 'any sockaddr_in6::sin6_addr that is not an "IPv4 address mapped into an IPv6 address" (i.e. any sockaddr_in6::sin6_addr for which IN6_IS_ADDR_V4MAPPED() is false)'.

Reply:
We have three "magic" labels for Acl::Ip
 - `all` matches the entire IPv4 and IPv6 Internet `::/0` range
 - `ipv4` matches all addresses in the `IN6_IS_ADDR_V4MAPPED()`. (because we store as IPv6 in that range)
 - `ipv6` matches all addresses that do not have a valid mapping to IPv4.

Note that there are multiple ranges that store IPv4 inside IPv6 space. Only one matches `ipv4` and none match `ipv6`. Ensuring that range gets handled correctly on input is left for a followup PR doing automatic re-maps.

Reply:
> * `ipv4` matches all addresses in the `IN6_IS_ADDR_V4MAPPED()`.
> * `ipv6` matches all addresses that do not have a valid mapping to IPv4.

Aha, I think we are making progress towards properly defining "ipv6" meaning! The discussion has exposed the existence of sockaddr_in6::sin6_addr values (i.e. in Squid context, "IP addresses") that are said to "have a valid mapping to IPv4" but for which IN6_IS_ADDR_V4MAPPED() is false (and, hence, ipv4 does not match). 

Would it be accurate to say that these IP addresses could be, in theory, mapped to IPv4, but were not mapped to IPv4 when _storing_ them in sockaddr_in6::sin6_addr, and that if they were to be mapped, IN6_IS_ADDR_V4MAPPED() would return true for the sockaddr_in6::sin6_addr values resulting from that mapping?

If Squid is configured to allow all traffic, what would happen to a request that targets such an IP address (explicitly in the request-target or as a result of that request-target DNS resolution)? Will Squid (with default/reasonable ./configure options and inside a modern IPv6-capable environment) be able to forward such a request? If yes, would the corresponding origin server IP address (used by Squid) be suitable for returning in a response to a DNS A query?

> Note that there are multiple ranges that store IPv4 inside IPv6 space. Only one matches ipv4 and none match ipv6. 

Could you please give (in a GitHub comment) a specific example of an IP address range that contains IP addresses that, in this PR code, will match neither ipv4 nor ipv6 ACLs? Since there are multiple such ranges and all are probably highly unusual, please pick the one that Squid is more likely (in your opinion) to encounter. Having a specific example in mind may help advance this discussion. If that range is documented in some RFC, please mention that.

How can we justify a design where Squid may see an IP address that matches neither ipv6 nor ipv4? Clearly, many (and possibly most!) admins expect the two ACLs to cover any IP address (that Squid accepts/stores as an IP address). Why are we not matching those (seemingly natural) admin expectations in our ipv4/ipv6 ACL definitions?

> Ensuring that range gets handled correctly on input is left for a followup PR

We are an agreement that Squid input validation needs work, and that we should strive to keep that work outside _this_ PR. However, I think this PR is the right place to establish ipv4 and ipv6 ACLs definitions and agree on the principles behind them. This thread is dedicated to that critical (IMO) work.